### ISSUE TITLE
Order splitting problem

### ISSUE THREAD

**Issue by CostaMateus** [weDevsOfficial/dokan#1690]
Context:
In my marketplace, some sellers ship products from two to three different places. Placing products from these sellers in the cart, ends up generating two different shipping costs.

Problem:
When products from two or more sellers are placed in the cart. And two products, from the same seller, with two different freights.
When the order is divided into sub-orders, the sub-order that should have two freight ends up losing the second freight.

I believe the screenshots make it easier to understand.

![1 products v1 e v2](https://user-images.githubusercontent.com/19847297/188249270-46c838eb-5eb3-4bcb-b156-8bf40d7aacff.png)
![3 checkout v1 e v2](https://user-images.githubusercontent.com/19847297/188249274-83f1194a-55d7-4cff-86b6-8e52c5d9e8a3.png)
![4 thankyou v1 e v2](https://user-images.githubusercontent.com/19847297/188249275-e038713d-955e-436a-b800-d0f03e9b895c.png)
![5 suborder 2 v1 e v2](https://user-images.githubusercontent.com/19847297/188249277-8da06064-9a60-4691-b16b-b1c26c0dd547.png)
![6 suborder 1 v1 e v2](https://user-images.githubusercontent.com/19847297/188249278-2414ff66-6d4e-4171-b0da-78eeb5b20893.png)

 - Dokan Version:  3.6.5
 - Dokan Pro Version:  3.7.5
 - WooCommerce Version: 6.8.2
 - Browser chrome
 - Current Theme: MartFury v2.8.3

**Comment by CostaMateus** [comment]
Looking at the code where the order is split. I made some changes to the create_shipping() method. Apparently I managed to solve the problem. However, I don't know if it causes side problems.

dokan-lite/includes/Order/Manager.php

original code:
`
    public function create_shipping( $order, $parent_order ) {
        dokan_log( sprintf( '#%d - Creating Shipping.', $order->get_id() ) );

        // Get all shipping methods for parent order
        $shipping_methods = $parent_order->get_shipping_methods();
        $order_seller_id  = dokan_get_seller_id_by_order( $order->get_id() );

        $applied_shipping_method = '';

        if ( $shipping_methods ) {
            foreach ( $shipping_methods as $method_item_id => $shipping_object ) {
                $shipping_seller_id = wc_get_order_item_meta( $method_item_id, 'seller_id', true );

                if ( $order_seller_id == $shipping_seller_id ) {
                    $applied_shipping_method = $shipping_object;
                    break;
                }
            }
        }

        $shipping_method = apply_filters( 'dokan_shipping_method', $applied_shipping_method, $order->get_id(), $parent_order );

        // bail out if no shipping methods found
        if ( ! $shipping_method ) {
            dokan_log( sprintf( '#%d - No shipping method found. Aborting.', $order->get_id() ) );
            return;
        }

        if ( is_a( $shipping_method, 'WC_Order_Item_Shipping' ) ) {
            $item = new \WC_Order_Item_Shipping();

            dokan_log( sprintf( '#%d - Adding shipping item.', $order->get_id() ) );

            $item->set_props(
                array(
                    'method_title' => $shipping_method->get_name(),
                    'method_id'    => $shipping_method->get_method_id(),
                    'total'        => $shipping_method->get_total(),
                    'taxes'        => $shipping_method->get_taxes(),
                )
            );

            $metadata = $shipping_method->get_meta_data();

            if ( $metadata ) {
                foreach ( $metadata as $meta ) {
                    $item->add_meta_data( $meta->key, $meta->value );
                }
            }

            $order->add_item( $item );
            $order->set_shipping_total( $shipping_method->get_total() );
            $order->save();
        }
    }
`

code with my changes:
`
    public function create_shipping( $order, $parent_order ) {
        dokan_log( sprintf( '#%d - Creating Shipping.', $order->get_id() ) );

        // Get all shipping methods for parent order
        $shipping_methods = $parent_order->get_shipping_methods();
        $order_seller_id  = dokan_get_seller_id_by_order( $order->get_id() );

        // $applied_shipping_method = '';
        // var changed to array
        $applied_shipping_methods = [];

        if ( $shipping_methods ) {
            foreach ( $shipping_methods as $method_item_id => $shipping_object ) {
                $shipping_seller_id = wc_get_order_item_meta( $method_item_id, 'seller_id', true );

                if ( $order_seller_id == $shipping_seller_id ) {
                    // $applied_shipping_method = $shipping_object;
                    // break;
                    
                    // dont break, checking all shipping methods
                    $applied_shipping_methods[] = $shipping_object;
                }
            }
        }

        // $shipping_method = apply_filters( 'dokan_shipping_method', $applied_shipping_method, $order->get_id(), $parent_order );
        // plural
        $shipping_methods = apply_filters( 'dokan_shipping_method', $applied_shipping_methods, $order->get_id(), $parent_order );

        // bail out if no shipping methods found
        // fix var
        if ( ! $shipping_methods ) {
            dokan_log( sprintf( '#%d - No shipping method found. Aborting.', $order->get_id() ) );
            return;
        }

        // add foreach
        foreach ( $shipping_methods as $shipping_method )
        {
            if ( is_a( $shipping_method, 'WC_Order_Item_Shipping' ) ) {
                $item = new \WC_Order_Item_Shipping();

                dokan_log( sprintf( '#%d - Adding shipping item.', $order->get_id() ) );

                $item->set_props(
                    array(
                        'method_title' => $shipping_method->get_name(),
                        'method_id'    => $shipping_method->get_method_id(),
                        'total'        => $shipping_method->get_total(),
                        'taxes'        => $shipping_method->get_taxes(),
                    )
                );

                $metadata = $shipping_method->get_meta_data();

                if ( $metadata ) {
                    foreach ( $metadata as $meta ) {
                        $item->add_meta_data( $meta->key, $meta->value );
                    }
                }

                $order->add_item( $item );
                $order->set_shipping_total( $shipping_method->get_total() );
                $order->save();
            }
        }
    }
`


Is it possible to confirm that there are no collateral problems?

**Comment by CostaMateus** [comment]
Any help?

**Comment by shohag121** [comment]
Hello @CostaMateus 

The code you gave as an example can be used with some modifications.
And the tax calculation also needs some modification as well.

We have decided to bring this change in **Dokan** natively. It will be added in the next version.

Thank you.

**Comment by CostaMateus** [comment]
Hello @shohag121 

Thanks for the feedback

**Comment by nurul-umbhiya** [comment]
Hello, @CostaMateus Are you using any third-party plugins to enable the per-product shipping feature? We couldn't produce the scenario on our end. If you explain the steps to reproduce the scenario that would be helpful for us. We are waiting for your reply.

**Comment by CostaMateus** [comment]
Hi @nurul-umbhiya.

I use the Claudio Sanches plugin, "Correios for WooCommerce", to add shipping methods to my marketplace.
In addition to that, I use a plugin I developed to integrate with a particular seller (FS).

In the marketplace, I only have one shipping zone set up. For normal sellers, for example (AA and BB), I have two shipping methods that are added by "Correios for WooCommerce". From this plugin I only use the methods: "SEDEX" and "PAC". The AA seller has the "SEDEX AA" and "PAC AA" methods. And the seller BB, the same thing. But for every normal seller I have only one shipping class set up. Both shipping methods use the same shipping class. All products from these sellers receive the same respective shipping class.

In the plugin I developed to integrate with the FS seller, the configuration is reversed. That is, I have a shipping method called "FS Delivery". And in this shipping method, I have three shipping classes configured, namely: "Center 1", "Center 2" and "Center 3". When importing FS seller products, they are assigned one of these three shipping classes. Each of these classes, even if they are sent to the same zip code, have different cost values.

So when I arrive at checkout, with three products from the FS seller, each of a different class, I end up with three different shipping charges.

I believe the main point to reproduce this scenario is the configuration of the shipping method class that extends WC_Shipping_Method. In the instance_form_fields, I configured the "shipping_class_id" field with the "multiselect" type. This configuration allows the shipping method to have more than one associated shipping class.
